src/scraper.py

```diff
@@ -0,0 +1,102 @@
+from contextlib import contextmanager
+import functools
+import html
+import re
+import time
+from urllib import robotparser
+
+from bs4 import BeautifulSoup
+from selenium import webdriver
+from selenium.webdriver.common.keys import Keys
+from selenium.webdriver.firefox.options import Options
+from selenium.webdriver.support.ui import WebDriverWait
+from selenium.common.exceptions import NoSuchElementException
+
+AGENT_NAME = 'codeforosnabrueckbot'
+
+
+def eval_robots_txt(agent_name):
+    @functools.wraps
+    def inner(func):
+        def wrapper(*args, **kwargs):
+            parser = robotparser.RobotFileParser(url=kwargs['url'])
+            parser.read()
+
+            if parser.can_fetch(agent_name, kwargs['url']):
+                return func(*args, **kwargs)
+            else:
+                raise PermissionError(f'The robots.txt permitts the crawling of the site {kwargs["url"]}')
+        return wrapper
+    return inner
+
+
+def wait_for_ajax(driver):
+    wait = WebDriverWait(driver, 1)
+    try:
+        wait.until(lambda driver: driver.execute_script('return jQuery.active') == 0)
+        wait.until(lambda driver: driver.execute_script('return document.readyState') == 'complete')
+    except Exception as e:
+        pass
+
+
+@eval_robots_txt(AGENT_NAME)
+def get_details(driver, *, url=None):
+    driver.find_element_by_tag_name('body').send_keys(Keys.CONTROL + 't')
+    driver.get(url)
+    wait_for_ajax(driver)
+
+    total_capacity = driver.find_element_by_class_name('detail-total-capacity').text
+    free_capacity = driver.find_element_by_class_name('detail-free-capacity').text
+
+    print(f'{free_capacity} von {total_capacity} frei.')
+
+    driver.find_element_by_tag_name('body').send_keys(Keys.CONTROL + 'w')
+
+    return map(int, (free_capacity, total_capacity))
+
+
+@eval_robots_txt(AGENT_NAME)
+def get_general_info(driver, *, url=None):
+    driver.get(url)
+    wait_for_ajax(driver)
+    time.sleep(1)
+
+    page_source = driver.page_source
+
+    parking_ramps = re.search(pattern='var parkingRampData = (\{.*\});',
+                              string=page_source)
+
+    parking_ramps = eval(html.unescape(parking_ramps.group(1)))
+
+    for identifier, ramp_data in parking_ramps.items():
+        print(f'Parking Ramp Name: {ramp_data["name"]}')
+
+        soup = BeautifulSoup(ramp_data['gmapsMarker'], 'html.parser')
+        details_url = soup.find('a', 'opg-map-infowindow-detaillink').get('href').replace('\\', '')
+
+        free_capacity, total_capacity = get_details(driver=driver, url=details_url)
+        ramp_data['free_capacity'] = free_capacity
+        ramp_data['total_capacity'] = total_capacity
+
+    return parking_ramps
+
+
+def main(url):
```

Comment:
I understand what you mean.
On the other hand, as Pythonista, I am used to read the bottom of a module right after the import statements because most often the ```__main__``` is located here. Therefore I would also expect an explicit but non mandatory ```main``` function to be located on the bottom.